lib/inc/pxp-agent/action_response.hpp

```diff
@@ -0,0 +1,82 @@
+#ifndef SRC_AGENT_ACTION_RESPONSE_HPP
+#define SRC_AGENT_ACTION_RESPONSE_HPP
+
+#include <pxp-agent/action_output.hpp>
+#include <pxp-agent/action_request.hpp>
+#include <pxp-agent/action_status.hpp>
+#include <pxp-agent/module_type.hpp>
+#include <pxp-agent/request_type.hpp>
+
+#include <leatherman/json_container/json_container.hpp>
+
+#include <string>
+#include <stdexcept>
+
+namespace PXPAgent {
+
+class ActionResponse {
+  public:
+    struct Error : public std::runtime_error {
+        explicit Error(std::string const& msg) : std::runtime_error(msg) {}
+    };
+
+    enum class ResponseType { Blocking, NonBlocking, StatusOutput, RPCError };
+
+    ModuleType module_type;
+    RequestType request_type;
+    ActionOutput output;
+    leatherman::json_container::JsonContainer action_metadata;
+
+    static leatherman::json_container::JsonContainer
+    getMetadataFromRequest(const ActionRequest& request);
+
+    ActionResponse(ModuleType module_type_,
+                   const ActionRequest& request_);
+
+    // Throws an Error if action_metadata is invalid.
+    ActionResponse(ModuleType module_type_,
+                   RequestType request_type_,
+                   ActionOutput output_,
+                   leatherman::json_container::JsonContainer&& action_metadata_);
+
+    void setStatus(ActionStatus status);
+
+    // Set results_are_valid to true in the action metadata and update
+    // the specified entries.
+    void setValidResultsAndEnd(leatherman::json_container::JsonContainer&& results,
+                               const std::string& execution_error = "");
+
+    // Set results_are_valid to false in the action metadata and
+    // update the execution_error entry.
+    void setBadResultsAndEnd(const std::string& execution_error);
```

Comment:
On the first reading the ```End``` in the names of the two methods above felt like a verb - _to end_, for that reason I'd prefer the method names to be: ```setExecutionErrorAndEndTimestamp``` and ```setExecutionErrorAndEndTimestamp``` respectively.